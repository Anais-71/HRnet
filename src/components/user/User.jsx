import React from 'react'
import PropTypes from 'prop-types'
import './user.css'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import {
  faXmark,
  faUser,
  faLocationDot,
  faBriefcase,
} from '@fortawesome/free-solid-svg-icons'

/**
 * User component that displays detailed information of a selected user.
 *
 * @component
 * @param {Object} props - The properties passed to the component.
 * @param {Object} props.user - The user object containing details to display.
 * @param {function} props.onClose - The function to call when closing the popup.
 *
 * @returns {JSX.Element} The rendered User details popup.
 */
const User = ({ user, onClose }) => (
  <div className="user">
    <div
      className="user__popup"
      role="dialog"
      aria-labelledby="userDetailsTitle"
      aria-describedby="userDetailsDesc"
    >
      <div className="user__popup--header">
        <h2 id="userDetailsTitle">User Details</h2>
        <button
          className="user__popup--close"
          onClick={onClose}
          aria-label="Close details"
          role="button"
        >
          <FontAwesomeIcon
            icon={faXmark}
            className="user__popup--close--icon"
          />
        </button>
      </div>

      <div
        className="user__popup--section"
        aria-labelledby="detailsSectionTitle"
        aria-describedby="detailsSectionDesc"
      >
        <div className="user__popup--section--header">
          <div className="user__popup--section--header--icon icon">
            <FontAwesomeIcon icon={faUser} className="icon__fa" />
          </div>
          <h3
            id="detailsSectionTitle"
            className="user__popup--section--header--title"
          >
            Details
          </h3>
        </div>
        <p id="detailsSectionDesc" className="user__popup--section--content">
          <strong>First Name:</strong> {user.firstName}
        </p>
        <p className="user__popup--section--content">
          <strong>Last Name:</strong> {user.lastName}
        </p>
        <p className="user__popup--section--content">
          <strong>Date of Birth:</strong> {user.dateOfBirth}
        </p>
      </div>

      <div
        className="user__popup--section"
        aria-labelledby="addressSectionTitle"
        aria-describedby="addressSectionDesc"
      >
        <div className="user__popup--section--header">
          <div className="user__popup--section--header--icon icon">
            <FontAwesomeIcon icon={faLocationDot} className="icon__fa" />
          </div>
          <h3
            id="addressSectionTitle"
            className="user__popup--section--header--title"
          >
            Address
          </h3>
        </div>
        <p id="addressSectionDesc" className="user__popup--section--content">
          <strong>Street:</strong> {user.street}
        </p>
        <p className="user__popup--section--content">
          <strong>City:</strong> {user.city}
        </p>
        <p className="user__popup--section--content">
          <strong>State:</strong> {user.state}
        </p>
        <p className="user__popup--section--content">
          <strong>Zip Code:</strong> {user.zipCode}
        </p>
      </div>

      <div
        className="user__popup--section"
        aria-labelledby="positionSectionTitle"
        aria-describedby="positionSectionDesc"
      >
        <div className="user__popup--section--header">
          <div className="user__popup--section--header--icon icon">
            <FontAwesomeIcon icon={faBriefcase} className="icon__fa" />
          </div>
          <h3
            id="positionSectionTitle"
            className="user__popup--section--header--title"
          >
            Position
          </h3>
        </div>
        <p id="positionSectionDesc" className="user__popup--section--content">
          <strong>Department:</strong> {user.department}
        </p>
        <p className="user__popup--section--content">
          <strong>Start Date:</strong> {user.startDate}
        </p>
      </div>
    </div>
  </div>
)

User.propTypes = {
  /**
   * The user object containing the details to display.
   * The shape of the user object is defined below.
   */
  user: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    dateOfBirth: PropTypes.string.isRequired,
    street: PropTypes.string.isRequired,
    city: PropTypes.string.isRequired,
    state: PropTypes.string.isRequired,
    zipCode: PropTypes.string.isRequired,
    department: PropTypes.string.isRequired,
    startDate: PropTypes.string.isRequired,
  }).isRequired,

  /**
   * The function to be called when the user clicks the close button.
   * This will typically be used to close the details popup.
   */
  onClose: PropTypes.func.isRequired,
}

export default User
