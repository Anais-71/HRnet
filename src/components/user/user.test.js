import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'
import React from 'react'
import User from './User'

const user = {
  firstName: 'Alice',
  lastName: 'Johnson',
  dateOfBirth: '1990-05-01',
  street: '123 Main St',
  city: 'Wonderland',
  state: 'CA',
  zipCode: '90001',
  department: 'Engineering',
  startDate: '2020-01-15',
}

/**
 * Test the rendering of the User component.
 * It should display the user details correctly.
 */
test('When I render the User component, then it displays the user details correctly', () => {
  render(<User user={user} onClose={() => {}} />)

  expect(screen.getByText('First Name:')).toBeInTheDocument()
  expect(screen.getByText('Alice')).toBeInTheDocument()

  expect(screen.getByText('Last Name:')).toBeInTheDocument()
  expect(screen.getByText('Johnson')).toBeInTheDocument()

  expect(screen.getByText('Date of Birth:')).toBeInTheDocument()
  expect(screen.getByText('1990-05-01')).toBeInTheDocument()

  expect(screen.getByText('Street:')).toBeInTheDocument()
  expect(screen.getByText('123 Main St')).toBeInTheDocument()

  expect(screen.getByText('City:')).toBeInTheDocument()
  expect(screen.getByText('Wonderland')).toBeInTheDocument()

  expect(screen.getByText('State:')).toBeInTheDocument()
  expect(screen.getByText('CA')).toBeInTheDocument()

  expect(screen.getByText('Zip Code:')).toBeInTheDocument()
  expect(screen.getByText('90001')).toBeInTheDocument()

  expect(screen.getByText('Department:')).toBeInTheDocument()
  expect(screen.getByText('Engineering')).toBeInTheDocument()

  expect(screen.getByText('Start Date:')).toBeInTheDocument()
  expect(screen.getByText('2020-01-15')).toBeInTheDocument()
})

/**
 * Test the close button functionality in the User component.
 * The close button should trigger the `onClose` function when clicked.
 */
test('When I click the close button, then the onClose function is called', () => {
  const onCloseMock = jest.fn()
  render(<User user={user} onClose={onCloseMock} />)

  const closeButton = screen.getByLabelText('Close details')

  // Simulate clicking the close button
  fireEvent.click(closeButton)

  // Verify that the onClose function is called once
  expect(onCloseMock).toHaveBeenCalledTimes(1)
})

/**
 * Test that the component has the appropriate ARIA attributes for accessibility.
 * The dialog should be labeled and described correctly with ARIA attributes.
 */
test('The User component has the correct ARIA attributes', () => {
  render(<User user={user} onClose={() => {}} />)

  const dialog = screen.getByRole('dialog')
  expect(dialog).toHaveAttribute('aria-labelledby', 'userDetailsTitle')
  expect(dialog).toHaveAttribute('aria-describedby', 'userDetailsDesc')

  const closeButton = screen.getByLabelText('Close details')
  expect(closeButton).toBeInTheDocument()
})

/**
 * Snapshot test for the User component.
 * The rendered output should match the saved snapshot.
 */
test('User component matches the snapshot', () => {
  const { asFragment } = render(<User user={user} onClose={() => {}} />)

  // Compare the rendered output to the saved snapshot
  expect(asFragment()).toMatchSnapshot()
})
