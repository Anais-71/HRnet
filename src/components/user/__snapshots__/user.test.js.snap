// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="user"
  >
    <div
      aria-describedby="userDetailsDesc"
      aria-labelledby="userDetailsTitle"
      class="user__popup"
      role="dialog"
    >
      <div
        class="user__popup--header"
      >
        <h2
          id="userDetailsTitle"
        >
          User Details
        </h2>
        <button
          aria-label="Close details"
          class="user__popup--close"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-xmark user__popup--close--icon"
            data-icon="xmark"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        aria-describedby="detailsSectionDesc"
        aria-labelledby="detailsSectionTitle"
        class="user__popup--section"
      >
        <div
          class="user__popup--section--header"
        >
          <div
            class="user__popup--section--header--icon icon"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-user icon__fa"
              data-icon="user"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512l388.6 0c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304l-91.4 0z"
                fill="currentColor"
              />
            </svg>
          </div>
          <h3
            class="user__popup--section--header--title"
            id="detailsSectionTitle"
          >
            Details
          </h3>
        </div>
        <p
          class="user__popup--section--content"
          id="detailsSectionDesc"
        >
          <strong>
            First Name:
          </strong>
           Alice
        </p>
        <p
          class="user__popup--section--content"
        >
          <strong>
            Last Name:
          </strong>
           Johnson
        </p>
        <p
          class="user__popup--section--content"
        >
          <strong>
            Date of Birth:
          </strong>
           1990-05-01
        </p>
      </div>
      <div
        aria-describedby="addressSectionDesc"
        aria-labelledby="addressSectionTitle"
        class="user__popup--section"
      >
        <div
          class="user__popup--section--header"
        >
          <div
            class="user__popup--section--header--icon icon"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-location-dot icon__fa"
              data-icon="location-dot"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 384 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
                fill="currentColor"
              />
            </svg>
          </div>
          <h3
            class="user__popup--section--header--title"
            id="addressSectionTitle"
          >
            Address
          </h3>
        </div>
        <p
          class="user__popup--section--content"
          id="addressSectionDesc"
        >
          <strong>
            Street:
          </strong>
           123 Main St
        </p>
        <p
          class="user__popup--section--content"
        >
          <strong>
            City:
          </strong>
           Wonderland
        </p>
        <p
          class="user__popup--section--content"
        >
          <strong>
            State:
          </strong>
           CA
        </p>
        <p
          class="user__popup--section--content"
        >
          <strong>
            Zip Code:
          </strong>
           90001
        </p>
      </div>
      <div
        aria-describedby="positionSectionDesc"
        aria-labelledby="positionSectionTitle"
        class="user__popup--section"
      >
        <div
          class="user__popup--section--header"
        >
          <div
            class="user__popup--section--header--icon icon"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-briefcase icon__fa"
              data-icon="briefcase"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M184 48l144 0c4.4 0 8 3.6 8 8l0 40L176 96l0-40c0-4.4 3.6-8 8-8zm-56 8l0 40L64 96C28.7 96 0 124.7 0 160l0 96 192 0 128 0 192 0 0-96c0-35.3-28.7-64-64-64l-64 0 0-40c0-30.9-25.1-56-56-56L184 0c-30.9 0-56 25.1-56 56zM512 288l-192 0 0 32c0 17.7-14.3 32-32 32l-64 0c-17.7 0-32-14.3-32-32l0-32L0 288 0 416c0 35.3 28.7 64 64 64l384 0c35.3 0 64-28.7 64-64l0-128z"
                fill="currentColor"
              />
            </svg>
          </div>
          <h3
            class="user__popup--section--header--title"
            id="positionSectionTitle"
          >
            Position
          </h3>
        </div>
        <p
          class="user__popup--section--content"
          id="positionSectionDesc"
        >
          <strong>
            Department:
          </strong>
           Engineering
        </p>
        <p
          class="user__popup--section--content"
        >
          <strong>
            Start Date:
          </strong>
           2020-01-15
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
